src/frameworks/karma/config/karma-config-loader.ts

```diff
@@ -67,10 +65,18 @@ export class KarmaConfigLoader {
     } else {
       const debugPortString = process.env[KarmaEnvironmentVariable.DebugPort];
       const debugPort: number | undefined = debugPortString ? parseInt(debugPortString, 10) : undefined;
+      this.logger.debug(() => `Using debug port: ${debugPort}`);
 
       const customLauncherString = process.env[KarmaEnvironmentVariable.CustomLauncher]!;
-      const customLaucherObject = customLauncherString ? JSON.parse(customLauncherString) : {};
-      this.addCustomLauncherDebugPort(customLaucherObject, debugPort);
+      let customLaucherObject = customLauncherString ? JSON.parse(customLauncherString) : {};
+
+      const browserHelper = BrowserHelperFactory.getBrowserHelper(customLaucherObject?.base ?? '');
+
+      if (!userSpecifiedLaunchConfig) {
+        customLaucherObject = this.findMatchingCustomLauncherFromConfig(customLaucherObject, config);
+      }
```

Comment:
Since it no longer tries to process the karma config file on the extension side, it's much more challenging to pull out all of the custom launcher configuration. In order to make use of the user's launch config it needs to figure out if one is being used here. This puts an existing launch config slightly higher priority than the generated launch config as the user's config can be more specific. If the user specifies a browser or defines custom launch config in the extension settings, this step will be ignored. It was the best solution I could come up with at the time that achieved the goal of allowing custom launch configuration from the karma config file to be used by the extension.

Reply:
To use a custom launch defined in the Karma config, I imagine you only need the name of that custom launcher, which you can place in the `browsers` prop of the config on the Karma process side. It seems your implementation already made a decent effort to obtain that from the extension process side of things and I'm pretty happy with the best effort that it makes there.

Reply:
It checks if one of the custom launchers uses a supported browser and then pulls out that browser to setup debug configuration. Could make it pass through the custom launcher name as if it were the browser setting, but then that doesn't support the debug port assignments if you have multiple projects active. Would still ignore the headless/non-headless browser setup if that was configured, assuming you take the browser name first.